clients/java/src/main/java/io/zeebe/client/impl/workflow/DeployWorkflowCommandImpl.java

```diff
@@ -0,0 +1,182 @@
+  }
+
+  @Override
+  public DeployWorkflowCommandBuilderStep2 addResourceStream(
+      final InputStream resourceStream, final String resourceName) {
+    ensureNotNull("resource stream", resourceStream);
+
+    try {
+      final byte[] bytes = StreamUtil.read(resourceStream);
+
+      return addResourceBytes(bytes, resourceName);
+    } catch (final IOException e) {
+      final String exceptionMsg =
+          String.format("Cannot deploy bpmn resource from stream. %s", e.getMessage());
+      throw new ClientException(exceptionMsg, e);
+    }
+  }
+
+  @Override
+  public DeployWorkflowCommandBuilderStep2 addResourceFromClasspath(
+      final String classpathResource) {
+    ensureNotNull("classpath resource", classpathResource);
+
+    try (final InputStream resourceStream =
+        getClass().getClassLoader().getResourceAsStream(classpathResource)) {
+      if (resourceStream != null) {
+        return addResourceStream(resourceStream, classpathResource);
+      } else {
+        throw new FileNotFoundException(classpathResource);
+      }
+
+    } catch (final IOException e) {
+      final String exceptionMsg =
+          String.format("Cannot deploy resource from classpath. %s", e.getMessage());
+      throw new RuntimeException(exceptionMsg, e);
+    }
+  }
+
+  @Override
+  public DeployWorkflowCommandBuilderStep2 addResourceFile(final String filename) {
+    ensureNotNull("filename", filename);
+
+    try (final InputStream resourceStream = new FileInputStream(filename)) {
+      return addResourceStream(resourceStream, filename);
+    } catch (final IOException e) {
+      final String exceptionMsg =
+          String.format("Cannot deploy resource from file. %s", e.getMessage());
+      throw new RuntimeException(exceptionMsg, e);
+    }
+  }
+
+  @Override
+  public DeployWorkflowCommandBuilderStep2 addWorkflowModel(
+      final BpmnModelInstance workflowDefinition, final String resourceName) {
+    ensureNotNull("workflow model", workflowDefinition);
+
+    final ByteArrayOutputStream outStream = new ByteArrayOutputStream();
+    Bpmn.writeModelToStream(outStream, workflowDefinition);
+    return addResourceBytes(outStream.toByteArray(), resourceName);
+  }
+
+  private ResourceType getResourceType(String resourceName) {
+    resourceName = resourceName.toLowerCase();
+
+    if (resourceName.endsWith(".yaml")) {
+      return YAML_WORKFLOW;
+    } else if (resourceName.endsWith(".bpmn") || resourceName.endsWith(".bpmn20.xml")) {
+      return BPMN_XML;
+    } else {
+      throw new RuntimeException(
+          String.format("Cannot resolve type of resource '%s'.", resourceName));
+    }
+  }
+
+  @Override
+  public ZeebeFuture<DeploymentEvent> send() {
+    final DeployWorkflowRequest.Builder request = DeployWorkflowRequest.newBuilder();
+
+    for (final DeploymentResource resource : resources) {
+      WorkflowRequestObject.ResourceType type = WorkflowRequestObject.ResourceType.BPMN;
```

Comment:
Why we need two enums for resource type? Could we remove the type in the client api? If yes then we only need one check and assign the type on the request building. If not then I would like to see something like the following instead of again an conditional statement.

```
WorkflowRequestObject.ResourceType type = WorkflowRequestObject.ResourceType.forNumber(resource.getResourceType().ordinal());
```

Reply:
Maybe it is also easier to use only the builder instead of extra properties for the resources in this class? Then we have no need to put them extra into the builder on send, we can call simply build.

Reply:
Technically, no we don't need them we can just stick the integer on the wire and call it a day. I wouldn't call it a nice design, but yes we could do it. What would we gain by getting rid of type safety on the wire?

Making them a hard dependency on each other makes things complicated for future changes.